docs: clarify locking mechanism requirement for S3

inigohidalgo opened the pull request:
- It was unclear to me that concurrent writing was available by default for non-S3 backends, so I am making the language clearer.
- I have also added an extra section showing that R2 and maybe MinIO can enable concurrent writing 
- Fixed a couple of unrelated formatting issues in the page I edited

closes #2556 

#2069 also had the same confusion

inigohidalgo commented:
> **ACTION NEEDED**
> 
> delta-rs follows the [Conventional Commits specification]

I have updated the pr title

ion-elgreco commented:
Can you mention the exception for Cloudflare R2, see this PR: https://github.com/delta-io/delta-rs/pull/2083

ion-elgreco commented:
https://github.com/delta-io/delta-rs/issues/1356#issuecomment-1736639674

inigohidalgo commented:
To make the docstrings explicit: what is the current status of other backends like MinIO? Can concurrent writes be enabled for them with DynamoDB? Or does that only apply to AWS?

ion-elgreco commented:
> To make the docstrings explicit: what is the current status of other backends like MinIO? Can concurrent writes be enabled for them with DynamoDB? Or does that only apply to AWS?

I'm not sure about Minio, @wjones127 mentions it supports custom headers so it might work as well for Minio like R2.

inigohidalgo commented:
Ok. Should I add a small section at the end (or beginning) of the S3 section with the example from the linked issue and also indicate that the same might be possible in MinIO? Or should I just mention the known R2 compatibility?

ion-elgreco commented:
You can mention that Minio might work in similar fashion, but needs to be double checked against their docs or something along those lines

inigohidalgo commented:
Just spotted this section in the readme https://github.com/delta-io/delta-rs/tree/b05d7e90dc0717b39c7fca35ec3c99c251aee839?tab=readme-ov-file#cloud-integrations

Should I change anything there for MinIO?

inigohidalgo commented:
I have added some cross-referencing links in the paragraphs I added so I wanted to check they work ok on my end before publishing the PR. I am having some trouble getting the docs to build but the rust + python docs builds are throwing me off a bit (I'm just used to Sphinx)

```bash
cd python
make develop  # crate and wheel build successfully
make build-docs  # installs all requirements correctly but then fails when building
```


```console
INFO    -  [macros] - Macros arguments: {'module_name': 'docs/_build/macro', 'modules': [], 'render_by_default': True, 'include_dir': '', 'include_yaml': [], 'j2_block_start_string': '', 'j2_block_end_string': '',
           'j2_variable_start_string': '', 'j2_variable_end_string': '', 'on_undefined': 'keep', 'on_error_fail': False, 'verbose': False}
INFO    -  [macros] - Found local Python module 'docs/_build/macro' in: /Users/inigo/Programming/Repos/delta-rs
INFO    -  [macros] - Found external Python module 'docs/_build/macro' in: /Users/inigo/Programming/Repos/delta-rs
INFO    -  [macros] - Extra variables (config file): ['python_api_url', 'generator', 'social']
INFO    -  [macros] - Extra filters (module): ['pretty']
INFO    -  Cleaning site directory
INFO    -  Building documentation to directory: /Users/inigo/Programming/Repos/delta-rs/site
INFO    -  mkdocstrings_handlers: Formatting signatures requires Black to be installed.
ERROR   -  Error reading page 'api/exceptions.md': Could not resolve alias deltalake._internal.DeltaError pointing at _internal.DeltaError (in python/deltalake/_internal.abi3.so:None)
Traceback (most recent call last):
  File "/Users/inigo/Programming/Repos/delta-rs/.venv/lib/python3.10/site-packages/griffe/dataclasses.py", line 1373, in _resolve_target
    resolved = self.modules_collection.get_member(self.target_path)
  File "/Users/inigo/Programming/Repos/delta-rs/.venv/lib/python3.10/site-packages/griffe/mixins.py", line 78, in get_member
    return self.members[parts[0]].get_member(parts[1:])  # type: ignore[attr-defined]
KeyError: '_internal'
```


If I open python I am able to access `deltalake.exceptions.DeltaError` without issue so I'm a bit lost on what the problem could be.

inigohidalgo commented on docs/usage/writing/writing-to-s3-with-locking-provider.md:
```diff
@@ -1,7 +1,10 @@
 
-A locking mechanism is needed to prevent unsafe concurrent writes to a
-delta lake directory when writing to S3.
+Delta lake guarantees [ACID transactions](https://delta-io.github.io/delta-rs/how-delta-lake-works/delta-lake-acid-transactions/) when writing data. This is done by default when writing to all supported object stores except AWS S3. (Some S3 clients like CloudFlare R2 or MinIO may enable atomic renames, refer to [this section](#enabling-concurrent-writes-for-alternative-clients) for more informati ...
```
I'd like to make this an internal link like sphinx's cross referencing, instead of just using the direct link, but I'm not sure what the correct syntax would be to point to the `delta-lake-acid-transactions` page from here

inigohidalgo commented on python/deltalake/writer.py:
```diff
@@ -208,9 +208,9 @@ def write_deltalake(
-    delta lake directory when writing to S3. For more information on the setup, follow
-    this usage guide: https://delta-io.github.io/delta-rs/usage/writing/writing-to-s3-with-locking-provider/
+    To enable safe concurrent writes when writing to S3, an additional locking
+    mechanism must be supplied. For more information on enabling concurrent writing to S3, follow  this usage guide: 
```
Is there a cross-referencing syntax I can use from within the docstring?

ion-elgreco replied:
Probably best to check mkdocs documentation

inigohidalgo reviewed (commented):
Cross-reference syntax

inigohidalgo commented:
I have verified the documentation builds on my end, and the links I added work okay.

ion-elgreco reviewed (approved):
Thankss for writing this up!

inigohidalgo commented:
https://github.com/delta-io/delta-rs/actions/runs/9332194054/job/25687819730?pr=2558

Oops. I forgot I had this same failure locally https://github.com/delta-io/delta-rs/pull/2558#issuecomment-2143366348, and I commented out the exceptions in `docs/api/exceptions.md`. I don't see how this can be related to my PR though.

inigohidalgo commented:
Okay I see that action has been failing for a while.
